### ISSUE TITLE
Python: Proper way of creating a separate thread in grpc server?

### ISSUE THREAD

**Issue by neliel123** [grpc/grpc#19119]
### What version of gRPC and what language are you using?
 1.11.0
 
### What operating system (Linux, Windows,...) and version?
 Linux
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 Python
 
### What did you do?
If possible, provide a recipe for reproducing the error. Try being specific and include code snippets if helpful.

I wanted to create a separate thread that will watch for changes in kubernetes server object and I wanted to save its values to some shared variables that I want to access in one of my rpc services.
I have written this looping thread using asyncio.
```python
async def watch_config_maps():
   v1 = client.CoreV1Api()
   is_watch = True
   w = watch.Watch()
   for event in w.stream(v1.list_namespaced_config_map,namespace="test-namespace"):
     #do something
     await asyncio.sleep(0)
		
def start_watcher():
  config.load_incluster_config()
  ioloop = asyncio.get_event_loop()
  ioloop.create_task(watch_config_maps())
  ioloop.run_forever()
```

And have it started together with my server.  
```python
def serve():
  global server
  server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))	
  #kube_watcher.start_watcher();
    
  server.add_insecure_port('[::]:' + str(port))
  server.start()
  try:
    while True:
      time.sleep(86400)
  except KeyboardInterrupt:
    server.stop(0)
```
But I am getting grpc unavailable message.  From this [issue](https://github.com/grpc/grpc/issues/19093) it says that Asyncio is not supported.
Now I would like to ask what should be the proper way of addresing my requirements.

Any ideas?
### What did you expect to see?
 I was hoping that I can gather information about events changes from the kubernetes server.

Basically, I want to create a looping thread where it will keep relevant information about Kubernetes Object.
When one of my RPC service gets called, I can look up at the values stored in my looping thread and
 
### What did you see instead?
 But when I tried running my asyncio method, all my grpc request is encountering error..
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with (StatusCode.UNAVAILABLE, Connect Failed)>

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.

**Comment by gnossen** [comment]
@neliel123 I may be missing some of the complexity of the dependencies you're pulling in, but I think this can be accomplished pretty simply using a standard Python `threading.Thread`. Something like:

```python
# The data to be shared between threads.
shared_map = {}
shared_map_lock = threading.RLock()

# The function to run in the background thread.
def watch_config_maps():
  v1 = client.CoreV1Api()
  is_watch = True
  w = watch.Watch()
  for event in w.stream(v1.list_namespaced_config_map,namespace="test-namespace"):
    key, value = process_event(event)
    with shared_map_lock:
      shared_map[key] = value

# Create the background thread and don't block interpreter exit on it.
thread = threading.Thread(target=watch_config_maps)
thread.daemon = True
thread.start()

# Pull from the shared data in the server handler method.
class Servicer:
  def handle_request(self, request, request_context):
    with shared_map_lock:
      return process_request(request, shared_map_lock)


```

**Comment by neliel123** [comment]
Thank you very much on this...
I was just picking up python grpc as I need to do something about it.

I am new to python as well so I am still working through it.  I tried this code and integrated it into my proof of concept project and they seem to work perfectly,

I do have a doubt on this line, with the locking on the lock..will I be encountering any performance hit on this?
If a handle_request is received and a lock is retrieved, will this block out other request?
```
# Pull from the shared data in the server handler method.
class Servicer:
  def handle_request(self, request, request_context):
    with shared_map_lock:
      return process_request(request, shared_map_lock)
```

Thanks for your time.  I very much appreciated it as a beginner working myself around python and python grpc as a whole.

**Comment by gnossen** [comment]
Sure. In general, you should hold the lock for as short a period of time as possible. I've updated my code sample above to reflect that.
